### PR TITLE
Fix/infinite product redirects

### PR DESCRIPTION
**Pull request by willPrattUPL**
## What/Why?
see https://github.com/bigcommerce/catalyst/issues/1320 for history

TLDR: Initial issue comment was incorrect. It seems to actually be a logical error when there are both a "redirect" and valid "node"

I am fairly certain there might be a "fix" upstream from this, but this fix resolves a showstopper bug (at least on our site)

We see this in an infinite loop:
```
. . . . [
@bigcommerce/catalyst-core:dev:   {
@bigcommerce/catalyst-core:dev:     "route": {
@bigcommerce/catalyst-core:dev:       "redirect": {
@bigcommerce/catalyst-core:dev:         "__typename": "Redirect",
@bigcommerce/catalyst-core:dev:         "to": {
@bigcommerce/catalyst-core:dev:           "__typename": "ProductRedirect"
@bigcommerce/catalyst-core:dev:         },
@bigcommerce/catalyst-core:dev:         "toUrl": "https://www.upliftdesk.com/uplift-v2-standing-desk-v2-or-v2-commercial/"
@bigcommerce/catalyst-core:dev:       },
@bigcommerce/catalyst-core:dev:       "node": {
@bigcommerce/catalyst-core:dev:         "__typename": "Product",
@bigcommerce/catalyst-core:dev:         "id": "UHJvZHVjdDo4OTA=",
@bigcommerce/catalyst-core:dev:         "entityId": 890
@bigcommerce/catalyst-core:dev:       }
@bigcommerce/catalyst-core:dev:     },
@bigcommerce/catalyst-core:dev:     "expiryTime": 1728411357384
@bigcommerce/catalyst-core:dev:   },
@bigcommerce/catalyst-core:dev:   {
@bigcommerce/catalyst-core:dev:     "status": "LAUNCHED",
@bigcommerce/catalyst-core:dev:     "expiryTime": 1728409857069
@bigcommerce/catalyst-core:dev:   }
@bigcommerce/catalyst-core:dev: ]
```

## Testing
I am unsure how to test without giving access to my forked copy / dev site. Please contact me for additional testing steps.

As far as I can tell, these are the criteria:

- Product route
- Has redirect
- Has valid node

## Context
The code fix is what we are using live, but our CLI install is about a month old and hasn't been upgraded since.

I have confirmed that this solution fixes the redirect loop on the latest version of cata on a local env, but I cannot confirm that the product page fully loads.

### CODE REVIEW

File: core/middlewares/with-routes.ts [hunk 1]
```diff
@@ -249,21 +249,26 @@ export const withRoutes: MiddlewareFactory = () => {
...
+    } else if (route?.redirect) {
       switch (route.redirect.to.__typename) {
         case 'ManualRedirect': {
           // For manual redirects, redirect to the full URL to handle cases
           // where the destination URL might be external to the site
           return NextResponse.redirect(route.redirect.toUrl, redirectConfig);
         }
-
+  
         default: {
```
Comment (chanceaclark) [inline core/middlewares/with-routes.ts:265]: Instead of adding additional logic before this switch statement, what if we just `break` on the `default` case and fallback to the logic below? It seems like if it's an internal redirect `node` will be populated which would fall into the normal `rewrite` logic 🤔 

This is what I mean https://github.com/bigcommerce/catalyst/compare/fix/redirects

Curious if that'll fix this issue?

Comment (willPrattUPL) [reply core/middlewares/with-routes.ts:265]: Can confirm this fixes the infinite redirect, but I don't know what side effect it has or how to test that accurately

Comment (chanceaclark) [reply core/middlewares/with-routes.ts:265]: @willPrattUPL can you list out what edge cases now return 404's when using the `break` method?

Comment (willPrattUPL) [reply core/middlewares/with-routes.ts:265]: with your early break implementation

## gets past infinite redirect & loads normally:



**/uplift-v2-standing-desk-v2-or-v2-commercial/**

Value: {
  "route": {
    "redirect": {
      "__typename": "Redirect",
      "to": {
        "__typename": "ProductRedirect"
      },
      "toUrl": "https://www.upliftdesk.com/uplift-v2-standing-desk-v2-or-v2-commercial/"
    },
    "node": {
      "__typename": "Product",
      "id": "UHJvZHVjdDo4OTA=",
      "entityId": 890
    }
  },
  "expiryTime": 1728479934882
}



**/uplift-v2-l-shaped-standing-desk/**

Value: {
  "route": {
    "redirect": {
      "__typename": "Redirect",
      "to": {
        "__typename": "ProductRedirect"
      },
      "toUrl": "https://www.upliftdesk.com/uplift-v2-l-shaped-standing-desk/"
    },
    "node": {
      "__typename": "Product",
      "id": "UHJvZHVjdDo4NTk=",
      "entityId": 859
    }
  },
  "expiryTime": 1728480227568
}




**/uplift-v2-standing-desk-frame/**

Value: {
  "route": {
    "redirect": {
      "__typename": "Redirect",
      "to": {
        "__typename": "ProductRedirect"
      },
      "toUrl": "https://www.upliftdesk.com/uplift-v2-standing-desk-frame/"
    },
    "node": {
      "__typename": "Product",
      "id": "UHJvZHVjdDoyNTQ=",
      "entityId": 254
    }
  },
  "expiryTime": 1728480292572
}


--------------------------------------------------------------------------------

## 404:

**/height-adjustable-standing-desk-with-l-shaped-top/**

[BigCommerce] KV MGET - Keys: [redact]/height-adjustable-standing-desk-with-l-shaped-top/,[redact] - Value: [
  {
    "route": {
      "redirect": {
        "__typename": "Redirect",
        "to": {
          "__typename": "ProductRedirect"
        },
        "toUrl": "https://www.upliftdesk.com/uplift-v2-l-shaped-standing-desk/"
      },
      "node": null
    },
    "expiryTime": 1728479977089
  },
  {
    "status": "LAUNCHED",
    "expiryTime": 1728478477014
  }
]
Debugger attached.
[BigCommerce] query RootLayoutMetadataQuery - 8ms - complexity 1006
[BigCommerce] query NotFoundQuery - 2ms - complexity 5400
[BigCommerce] query LayoutQuery - 186ms - complexity 2225
[BigCommerce] query RootLayoutMetadataQuery - 2ms - complexity 1006
 GET /height-adjustable-standing-desk-with-l-shaped-top/ 404 in 1857ms
Waiting for the debugger to disconnect...

Comment (chanceaclark) [reply core/middlewares/with-routes.ts:265]: Hmmm `node` coming back `null` for a `ProductRedirect` seems like a bug on our API 🤔 I'm going to check with the team if that's the case.

Comment (chanceaclark) [reply core/middlewares/with-routes.ts:265]: Hey @willPrattUPL just chatted with the team and saw we have an option on the `route` node that might fix that last 404 edge case. Try setting the `redirectBehavior` to `FOLLOW` in the query along with the `break;` statement flow to see if that fixes the issue.
Screenshot 2024-10-11 at 13 55 12

Comment (willPrattUPL) [reply core/middlewares/with-routes.ts:265]: So just to clarify, you are saying to add this to the query in get-route:

      route(path: $path, redirectBehavior: FOLLOW) {
      
      ?
      
      If so, that does indeed fix the edge case, as well

Comment (chanceaclark) [reply core/middlewares/with-routes.ts:265]: Yea! That's great then! I think we update this PR to that instead, that way we don't need to duplicate that similar logic.

**Comment by willPrattUPL** [comment]
Updated to suggested improvements

**Review by chanceaclark — APPROVED** [review]
Looks good to me! Added a changeset for this PR.
